snapcraft/cli/_options.py

```diff
@@ -65,49 +65,97 @@ def prompt_for_value(self, ctx):
+        option = option.copy()
+        param_decls = option.pop("param_decls")
+        click_option = click.option(param_decls, **option, hidden=hidden)
+        func = click_option(func)
+    return func
+
+
 def add_build_options(hidden=False):
     def _add_build_options(func):
-        for build_option in reversed(_BUILD_OPTIONS):
-            # Pop param_decls option to prevent exception further on down the
-            # line for: `got multiple values for keyword argument`.
-            build_option = build_option.copy()
-            param_decls = build_option.pop("param_decls")
-            option = click.option(param_decls, **build_option, hidden=hidden)
-            func = option(func)
-        return func
+        return _add_options(_BUILD_OPTIONS, func, hidden)
 
     return _add_build_options
 
 
+def add_provider_options(hidden=False):
+    def _add_provider_options(func):
+        return _add_options(_PROVIDER_OPTIONS, func, hidden)
+
+    return _add_provider_options
+
+
 def get_build_environment(**kwargs):
-    force_use_lxd = kwargs.get("use_lxd")
-    force_destructive_mode = kwargs.get("destructive_mode")
+    provider = kwargs.get("provider")
+    use_lxd = kwargs.get("use_lxd")
+    destructive_mode = kwargs.get("destructive_mode")
 
-    if force_destructive_mode and force_use_lxd:
+    if destructive_mode and use_lxd:
         raise click.BadOptionUsage(
             "--destructive-mode",
             "--use-lxd and --destructive-mode cannot be used together.",
         )
-    elif force_use_lxd:
-        provider = "lxd"
-    elif force_destructive_mode:
-        provider = "host"
+
+    if provider:
+        if provider not in ["lxd", "host", "multipass"]:
+            raise click.BadOptionUsage(
+                "--provider",
+                "`--provider` valid options are `lxd`, `host`, or `multipass`.",
+            )
+
+        if provider != "lxd" and use_lxd:
+            raise click.BadOptionUsage(
+                "--provider",
+                "--use-lxd cannot be used with `--provider={}`".format(provider),
+            )
+
+        if provider != "host" and destructive_mode:
+            raise click.BadOptionUsage(
+                "--provider",
+                "--destructive-mode cannot be used with `--provider={}`".format(
+                    provider
+                ),
+            )
+
+        if provider == "host" and not destructive_mode:
+            raise click.BadOptionUsage(
+                "--provider",
+                "`--provider=host` requires `--destructive-mode` to acknowledge side effects.",
+            )
     else:
-        provider = None
-    return env.BuilderEnvironmentConfig(force_provider=provider)
+        if use_lxd:
+            provider = "lxd"
+
+        if destructive_mode:
+            provider = "host"
+
+    # BuilderEnvironmentConfig will determine provider if None.
+    return env.BuilderEnvironmentConfig(build_provider=provider)
```

Comment:
getting rid of BuildEnvironmentConfig might also be fun :-)

Reply:
Done! And with that also managed to consolidate SNAPCRAFT_BUILD_ENVIRONMENT handling in with the --provider option. Much nicer :D